modules/loaders-common/src/main/scala/com/snowplowanalytics/snowplow/loaders/transform/NonAtomicFields.scala

```diff
@@ -47,18 +50,65 @@ object NonAtomicFields {
 
   def resolveTypes[F[_]: Sync: RegistryLookup](
     resolver: Resolver[F],
-    entities: Map[TabledEntity, Set[SchemaSubVersion]]
+    entities: Map[TabledEntity, Set[SchemaSubVersion]],
+    filterCriteria: List[SchemaCriterion]
   ): F[Result] =
     entities.toList
+      .filter { case (tabledEntity, subVersions) =>
```

Comment:
Did you consider something like...

```scala
entities.toList
  .map { case (tabledEntity, subVersions) =>
    (tabledEntity -> subVersions.filter(???)
  }
  .filter { case (_, subVersions) =>
    subVersions.nonEmpty
  }
  .traverse {
    // What was here originally
  }

Is there an edge case I haven't thought of?

Reply:
Even if we map + filter `subVersions` early,  `fetchSchemasWithSameMode` later returns all available.

Reply:
Let's dig into this.  The purpose of these lines of code....

```scala
SchemaProvider
  .fetchSchemasWithSameModel(resolver, TabledEntity.toSchemaKey(tabledEntity, subVersions.max))
  .map(TypedTabledEntity.build(tabledEntity, subVersions, _))
```

is to build a merged data type that takes into account all relevant schemas in a family.

For example, imagine a schema has four versions `1-0-0`, `1-0-1`, `1-0-2`, `1-0-3`.  And imagine this pipeline is currently receiving events for `1-0-0` and `1-0-2`.  These lines of code then fetch schemas `1-0-1`, `1-0-1`, and `1-0-2` from the Iglu Server, and merge the three schemas together.  If `1-0-2` contained a breaking change then the end result include a recovery schema, as well as a merged schema.

Now imagine the loader is configured with `skipSchemas: [ vendor/name/jsonschema/1-0-0 ]`.   With your current code change, we now consider only schemas `1-0-1` and `1-0-2` when merging the schemas.  But I think this is incorrect!  The result of merging two schemas might give a merge-result which is not compatible with the table.

In other words, I think when we call `TypedTabledEntity.build(...)` it is important we pass in a contiguous series of schemas.

---

**Suggestion 1**

How about, we use all schemas to build the typed entities, but then afterwards filter the `Result` to remove unwanted schema versions:

```scala
SchemaProvider
  .fetchSchemasWithSameModel(resolver, TabledEntity.toSchemaKey(tabledEntity, subVersions.max))
  .map(TypedTabledEntity.build(tabledEntity, subVersions, _))
  .map(tte => filterSubVersionsFromTTE(tte, filterCriteria))
```

**Suggestion 2**

How about we only support skipping whole families of schemas.  E.g. the schema criterion MUST be like `1-*-*` not `1-0-0`.

This means you'd need to amend `shouldWeKeepEntity` but everything below that can stay like it was originally.

The only small problem is... I don't think we have a nice Iglu case class for "schema key where revision is None and addition is None".

**Suggestion 3**

How about we make it a rule... if you skip version `1-0-2` then by implication we will also skip `1-0-3` and `1-0-4`.  It is not allowed to skip only `1-0-2`.

This would require a filtering step somewhere around `shouldWeKeepEntity` but everything below that can stay like it was originally.

**Suggestion 4**

Go with your original suggestion, where we apply the filter inside `TabledEntity.forEvent` instead of here.  I know I raised questions about efficiency -- but it's probably fine, and it avoids the issues under discussion here.